File: bench/matmul.c
// Dense integer matrix multiply, N x N, ikj order. Matches matmul.zeph/.rs.
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

int main(int argc, char** argv) {
    int N = argc > 1 ? atoi(argv[1]) : 512;
    int64_t* A = malloc((size_t)N * N * sizeof(int64_t));
    int64_t* B = malloc((size_t)N * N * sizeof(int64_t));
    int64_t* C = calloc((size_t)N * N, sizeof(int64_t));
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++) {
            A[i * N + j] = (i * 3 + j * 7 + 1) % 10;
            B[i * N + j] = (i * 5 + j * 2 + 3) % 10;
        }
    for (int i = 0; i < N; i++)
        for (int k = 0; k < N; k++) {
            int64_t aik = A[i * N + k];
            for (int j = 0; j < N; j++)
                C[i * N + j] += aik * B[k * N + j];
        }
    int64_t sum = 0;
    for (int i = 0; i < N * N; i++) sum += C[i];
    printf("%lld\n", (long long)sum);
    return 0;
}

File: bench/matmul.rs
// Dense integer matrix multiply, N x N, ikj order. Matches matmul.zeph/.c.
use std::env;

fn main() {
    let n: usize = env::args().nth(1).and_then(|a| a.parse().ok()).unwrap_or(512);
    let mut a = vec![0i64; n * n];
    let mut b = vec![0i64; n * n];
    let mut c = vec![0i64; n * n];
    for i in 0..n {
        for j in 0..n {
            a[i * n + j] = ((i * 3 + j * 7 + 1) % 10) as i64;
            b[i * n + j] = ((i * 5 + j * 2 + 3) % 10) as i64;
        }
    }
    for i in 0..n {
        for k in 0..n {
            let aik = a[i * n + k];
            for j in 0..n {
                c[i * n + j] += aik * b[k * n + j];
            }
        }
    }
    let sum: i64 = c.iter().sum();
    println!("{}", sum);
}
